### PR TITLE
Make the function call-rate limiting a generic thing and use it in more places.

### PR DESCRIPTION
**Pull request by dbkr**
Fixes https://github.com/vector-im/vector-web/issues/779

### CODE REVIEW

File: src/ratelimitedfunc.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+            f.apply(this);
+            self.lastCall = now;
+        } else if (self.scheduledCall === undefined) {
+            self.scheduledCall = setTimeout(() => {
+                    self.scheduledCall = undefined;
+                    f.apply(this);
+                    self.lastCall = now;
+                },
+                (self.lastCall + min_interval_ms) - now
+            );
```
Comment (Kegsay) [inline src/ratelimitedfunc.js]: Not sure about the indentation here. I was expecting to see a `}` on this line from the `setTimeout` function definition. It's not immediately obvious which things are open/closed with fat arrow syntax like `(() => {` , hence I would prefer:

```javascript
self.scheduledCall = setTimeout(() => {
    // stuff
}, (self.lastCall + min_interval_ms) - now);
```
or (but this is much more whitespace :( ):
```javascript
self.scheduledCall = setTimeout(
    () => {
        // stuff
    },
    (self.lastCall + min_interval_ms) - now
);
```

File: src/ratelimitedfunc.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+module.exports = function(f, min_interval_ms) {
+    this.lastCall = 0;
+    this.scheduledCall = undefined;
```
Comment (Kegsay) [inline src/ratelimitedfunc.js]: `camelCase` or `under_scores`, which is it? We don't really fix on a certain style in react/vector, but we should at least be consistent within a file (especially new files).

File: src/components/views/rooms/RoomList.js [hunk 3]
```diff
@@ -155,36 +156,9 @@ module.exports = React.createClass({
...
-                    self._refreshRoomListTimerId = setTimeout(function() {
-                        self.refreshRoomList();
-                    }, 10 + MAX_REFRESH_INTERVAL_MS - timeWaitedMs); // 10 is a buffer amount
-                }
-            }
-        }
-    },
+    _delayedRefreshRoomList: new rate_limited_func(function() {
+        this.refreshRoomList();
+    }, 500),
```
Comment (Kegsay) [inline src/components/views/rooms/RoomList.js]: We've dropped this from 1000 ms to 500 ms. Intentional?

**Comment by Kegsay** [comment]
LGTM other than little style things.

File: src/components/views/rooms/RoomList.js (hunk 3 above)
Comment (dbkr) [inline src/components/views/rooms/RoomList.js]: Yeah, I kept it at 1000 to try, and dropped it - it works just as well for me with 500 and 1000 seems a little long to delay updates.

**Comment by Kegsay** [comment]
LGTM

**Pull request merged**